src/Microsoft.Android.Runtime.NativeAOT/Android.Runtime.NativeAOT/TypeMapping.cs

```diff
@@ -0,0 +1,68 @@
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.IO.Hashing;
+using System.Runtime.InteropServices;
+using System.Text;
+
+using Android.Runtime;
+
+namespace Microsoft.Android.Runtime;
+
+internal static class TypeMapping
+{
+	// Initialization generated by src/Microsoft.Android.Sdk.ILLink/TypeMappingStep.cs
+	private static readonly ulong[] s_hashes = null!;
+	private static Dictionary<Type, string[]> s_javaClassNames = new ();
+
+	static TypeMapping ()
+	{
+		Initialize ();
+	}
+
+	internal static int Count => s_hashes.Length;
+
+	internal static bool TryGetType (string javaClassName, [NotNullWhen (true)] out Type? type)
+	{
+		ReadOnlySpan<byte> bytes = MemoryMarshal.AsBytes(javaClassName.AsSpan ());
+		ulong hash = XxHash3.HashToUInt64 (bytes);
+		type = GetType (hash);
+		return type is not null;
+	}
+
+	internal static IEnumerable<string> GetJavaClassNames (Type type)
+	{
+		if (!s_javaClassNames.TryGetValue (type, out var javaClassNames)) {
+			javaClassNames = GetJavaClassNamesSlow (type).ToArray ();
+			_ = s_javaClassNames.TryAdd (type, javaClassNames);
+		}
+
+		return javaClassNames;
+	}
+
+	private static IEnumerable<string> GetJavaClassNamesSlow (Type type)
+	{
+		for (int i = 0; i < Count; i++) {
+			if (GetTypeByIndex (i) == type) {
+				var javaClassName = TypeMapping.GetJavaClassNameByIndex (i);
+				if (javaClassName is not null) {
+					yield return javaClassName;
+				}
+			}
+		}
+	}
+
+	private static Type? GetType (ulong hash)
+	{
+		int index = MemoryExtensions.BinarySearch (s_hashes, hash);
```

Comment:
Does this assume that we will be lucky and there is never going to be a hashcode collision?

Reply:
Yes. We could compare `GetJavaClassNameByIndex(index)` to the input class name if we wanted to make sure there isn't a collision, but AFAIK we aren't doing these checks in the current native type maps used in .NET 9 and earlier, so I don't think it's a big concern. Please correct me if I'm wrong, @grendello.

Reply:
It depends on the quality/reliability bar.

FWIW, depending on non-colliding hashcodes like this would not be acceptable for dotnet/runtime.

Reply:
@simonrozsival you're not wrong. This is perfectly fine for a small set of strings we deal with, and a high quality hash like xxHash. The collision checks should be performed at build time, just in case, but at runtime we can safely assume there are no collisions.

Reply:
> The collision checks should be performed at build time

This is 100% reliable only if you can see the full set of inputs that this method is expected to handle at build time. Do you have a guarantee like that? In other words, is it an error if the method does find the class name and returns null?

Reply:
> This is 100% reliable only if you can see the full set of inputs that this method is expected to handle at build time. Do you have a guarantee like that?

We don't know all the possible Java class names at build time. I added the check that the Java class names match. We can return to this at some later point and see if this additional check even makes any meassurable impact on app startup times.